### DISCUSSION TITLE
VST 2.4

### DISCUSSION THREAD

**Discussion by rotemdoron** [apohl79/audiogridder#431] [Q&A] [answered]
Hi All, I'm having issue's building the project, I managed to build using CMake and generated an Xcode project with CMake but I'm getting an error when trying to build `'pluginterfaces/vst2.x/aeffect.h' file not found` 
It's looking for the VST 2.4 SDK but can't seem to find it anywhere as Steinberg dropped support, would love to know how did you guys manage to build the project.  Thank you! :)

**Comment by apohl79** [comment, marked as answer]
Try this link: https://archive.org/details/VST2SDK

**Reply by rotemdoron** [reply]
Hi, @apohl79 thanks for replying! looks like that fixed the VST 2.4 error, still can't build the project getting a lot of 
`Undefined symbols `(around 100+) error messages, can't figure out what's missing

**Reply by apohl79** [reply]
You need boost, libwebp and ffmpeg. Maybe thats whats missing. Try 

```brew install boost webp ffmpeg```

**Reply by rotemdoron** [reply]
Got them up and running actually, not sure what's the issue but I'm seeing in `Build Phases` -> `Link Binary With Libraries ` that there are no frameworks there, you think that might be the issue? I thought CMakeList takes care of linking, maybe im wrong. again thank you for replying back!

**Reply by apohl79** [reply]
Ok the problem is that AG links everything statically, and I use custom builds of webp and ffmpeg that contain just what I need. Now, when you use the brew versions of those libs, they have much more dependencies (which you would also have to link statically into AG) and thus you get unresolved symbols.

Here is how you can enable dynamic linking:
```
diff --git a/cmake/FindFFmpeg.cmake b/cmake/FindFFmpeg.cmake
index 83005cc..c7d8ca5 100644
--- a/cmake/FindFFmpeg.cmake
+++ b/cmake/FindFFmpeg.cmake
@@ -13,7 +13,7 @@ macro(FFMPEG_FIND varname shortname headername)
     DOC "Location of FFMPEG Headers")

   find_library(FFMPEG_${varname}_LIBRARIES
-    NAMES lib${shortname}.a lib${shortname}.lib
+    NAMES lib${shortname}.dylib lib${shortname}.lib
     PATHS
     ${FFMPEG_ROOT}/lib
     $ENV{FFMPEG_DIR}/lib
diff --git a/cmake/FindWebP.cmake b/cmake/FindWebP.cmake
index 7c5aafb..3a4735e 100644
--- a/cmake/FindWebP.cmake
+++ b/cmake/FindWebP.cmake
@@ -13,7 +13,7 @@ macro(WEBP_FIND varname shortname headername)
     DOC "Location of FFMPEG Headers")

   find_library(WEBP_${varname}_LIBRARIES
-    NAMES lib${shortname}.a lib${shortname}.lib
+    NAMES lib${shortname}.dylib lib${shortname}.lib
     PATHS
     ${FFMPEG_ROOT}/lib
     $ENV{FFMPEG_DIR}/lib
```

**Reply by rotemdoron** [reply]
Hi @apohl79 thanks for this! may I ask you where is this script suppose to go? CMakeList or Build Script in XCode Build Phases or something like that?

**Reply by rotemdoron** [reply]
I Guess it's `FindFFmpeg.cmake` and `FindWebP.cmake`

**Reply by rotemdoron** [reply]
tried it, still getting undefined symbols, there must be something I'm missing 

Showing All Messages
Undefined symbol: _D_IF_decode

Undefined symbol: _D_IF_exit

Undefined symbol: _D_IF_init

Undefined symbol: _Decoder_Interface_Decode

Undefined symbol: _Decoder_Interface_exit

Undefined symbol: _Decoder_Interface_init

Undefined symbol: _Encoder_Interface_Encode

Undefined symbol: _Encoder_Interface_exit

Undefined symbol: _Encoder_Interface_init

Undefined symbol: _VTCompressionSessionCompleteFrames

Undefined symbol: _VTCompressionSessionCreate

Undefined symbol: _VTCompressionSessionEncodeFrame

Undefined symbol: _VTCompressionSessionGetPixelBufferPool

Undefined symbol: _VTCompressionSessionPrepareToEncodeFrames

Undefined symbol: _VTDecompressionSessionCreate

Undefined symbol: _VTDecompressionSessionDecodeFrame

Undefined symbol: _VTDecompressionSessionInvalidate

Undefined symbol: _VTDecompressionSessionWaitForAsynchronousFrames

Undefined symbol: _VTSessionCopyProperty

Undefined symbol: _VTSessionSetProperty

**Reply by apohl79** [reply]
Just change the ffmpeg and webp cmake files. The changed lines make cmake look for the .dylib files rather than .a.

Does cmake build successfully and just xcode not? What macos version are you on?

How did you run cmake?

cmake -B build
cmake --build build 

This should be it. If that still does not work, attach the output please.

**Reply by rotemdoron** [reply]
So yeap Xcode is the issue, I run 
cmake -B build
cmake --build build
which runs but at the end of the process im getting:
AudioGridder: no identity found
make[2]: *** [Plugin/CMakeFiles/AudioGridderMidi_VST3_SIGN] Error 1
make[1]: *** [Plugin/CMakeFiles/AudioGridderMidi_VST3_SIGN.dir/all] Error 2
make: *** [all] Error 2

also tried to generate the Xcode project with cmake -G Xcode

**Reply by apohl79** [reply]
Ignore the signing errors. Signing happens after successful building. The binaries are located in the artifacts directories in the build dir.

**Reply by rotemdoron** [reply]
@apohl79 That's awesome thanks so much!🙏  it is working now! when I run the plugin, even though the build succeeds, i open for example Ableton but after running the FX, for example, it stops running right after the build so can't really debug it that way, maybe you have an idea

**Reply by apohl79** [reply]
If ableton allows it, you can run it in lldb via command line or xcode. From there you can debug the plugin.

**Reply by apohl79** [reply]
FYI, added a few more options to CMake so you can activate dynamic linking w/o patching the cmake files. Just pull from master.

**Reply by rotemdoron** [reply]
Hi @apohl79, thank you this is great!

**Reply by rotemdoron** [reply]
Hi @apohl79, seems i can't really build the plugin, I generated an Xcode project using `cmake -G Xcode` so i can debug but when I run for example the FX VST3 build scheme either on my mac or directly to Ableton as executable, nothing happens the build succeeds but then it stops and it doesn't install the plugin anywhere. would love your help maybe I got it wrong

**Reply by apohl79** [reply]
I'm not using Xcode. Also I have never really worked with cmakes xcode generator.

Regarding copying, when you are using the "Debug" build config (-DCMAKE_BUILD_TYPE=Debug) there is a COPY target for the binaries. This final cmake step is copying the binaries to ~/Library/Audio/Plug-Ins/... After this your DAW should find them.

Now you can run the DAW in lldb. (eg: lldb "/Applications/Logic Pro X.app/Contents/MacOS/Logic Pro X", put ableton here, run it and load an AG plugin etc)

The problem is probably, that the copy step is executed by cmake, not Xcode. You can try working with symlinks from the plugin folders into your build folder. You might also be able to extend the generated xcode project, but you need to figure this out yourself.
